File: backend/app/schemas/takes.py
"""
Pydantic schemas for takes system.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class GenerationParams(BaseModel):
    """Parameters for content generation"""

    model: str = Field(..., description="Model used for generation")
    seed: int = Field(..., description="Random seed for reproducibility")
    prompt: str = Field(..., description="Generation prompt")
    negative_prompt: str | None = Field(None, description="Negative prompt")
    steps: int = Field(..., description="Number of generation steps")
    cfg: float = Field(..., description="Classifier-free guidance scale")
    # Allow additional model-specific parameters
    extra: dict[str, Any] = Field(default_factory=dict, description="Additional parameters")


class Resolution(BaseModel):
    """Media resolution"""

    width: int = Field(..., description="Width in pixels")
    height: int = Field(..., description="Height in pixels")


class ResourceUsage(BaseModel):
    """Resource usage for generation"""

    quality: str = Field(..., description="Quality level: draft, standard, high")
    vram_used: float = Field(0, description="Peak VRAM usage in MB")
    generation_time: float = Field(0, description="Generation time in seconds")


class TakeMetadata(BaseModel):
    """Complete metadata for a take"""

    id: str = Field(..., description="Take ID (e.g., take_001)")
    shot_id: str = Field(..., description="Shot ID")
    created: datetime = Field(..., description="Creation timestamp")
    duration: float | None = Field(None, description="Video duration in seconds")
    resolution: Resolution | None = Field(None, description="Media resolution")
    generation_params: GenerationParams = Field(..., description="Generation parameters")
    resources: ResourceUsage = Field(..., description="Resource usage")
    status: str = Field(..., description="Status: generating, complete, failed")
    error: str | None = Field(None, description="Error message if failed")
    file_path: str | None = Field(None, description="Relative path to media file")
    file_size: int | None = Field(None, description="File size in bytes")
    thumbnail_path: str | None = Field(None, description="Relative path to thumbnail")


class CreateTakeRequest(BaseModel):
    """Request to create a new take"""

    generation_params: GenerationParams = Field(..., description="Generation parameters")
    quality: str = Field("standard", description="Quality level")


class CreateTakeResponse(BaseModel):
    """Response after creating a take"""

    take_id: str = Field(..., description="Generated take ID")
    take_number: int = Field(..., description="Take number")
    status: str = Field("queued", description="Initial status")
    task_id: str | None = Field(None, description="Task ID for tracking")


class TakeListResponse(BaseModel):
    """Response with list of takes"""

    takes: list[TakeMetadata] = Field(..., description="List of takes")
    active_take_id: str | None = Field(None, description="Currently active take")
    total: int = Field(..., description="Total number of takes")


class SetActiveTakeRequest(BaseModel):
    """Request to set active take"""

    take_id: str = Field(..., description="Take ID to make active")


class DeleteTakeResponse(BaseModel):
    """Response after deleting a take"""

    success: bool = Field(..., description="Whether deletion succeeded")
    message: str = Field(..., description="Status message")
    new_active_take_id: str | None = Field(
        None, description="New active take if the deleted one was active"
    )


class TakeExportRequest(BaseModel):
    """Request to export a take"""

    take_id: str = Field(..., description="Take ID to export")
    include_metadata: bool = Field(True, description="Include metadata file")


class TakeExportResponse(BaseModel):
    """Response after exporting a take"""

    success: bool = Field(..., description="Whether export succeeded")
    export_path: str | None = Field(None, description="Path to exported file")
    message: str = Field(..., description="Status message")


class TakeCleanupRequest(BaseModel):
    """Request to cleanup old takes"""

    keep_count: int = Field(10, description="Number of recent takes to keep", ge=1, le=100)
    include_failed: bool = Field(True, description="Also cleanup failed takes")


class TakeCleanupResponse(BaseModel):
    """Response after cleaning up takes"""

    deleted_count: int = Field(..., description="Number of takes deleted")
    remaining_count: int = Field(..., description="Number of takes remaining")
    message: str = Field(..., description="Status message")

File: frontend/src/lib/api/takes.ts
import { api } from './client';

export interface GenerationParams {
  model: string;
  seed: number;
  prompt: string;
  negativePrompt?: string;
  steps: number;
  cfg: number;
  extra?: Record<string, any>;
}

export interface Resolution {
  width: number;
  height: number;
}

export interface ResourceUsage {
  quality: string;
  vramUsed: number;
  generationTime: number;
}

export interface TakeMetadata {
  id: string;
  shotId: string;
  created: string;
  duration?: number;
  resolution?: Resolution;
  generationParams: GenerationParams;
  resources: ResourceUsage;
  status: 'generating' | 'complete' | 'failed';
  error?: string;
  filePath?: string;
  fileSize?: number;
  thumbnailPath?: string;
}

export interface CreateTakeRequest {
  generationParams: GenerationParams;
  quality?: string;
}

export interface CreateTakeResponse {
  takeId: string;
  takeNumber: number;
  status: string;
  taskId?: string;
}

export interface TakeListResponse {
  takes: TakeMetadata[];
  activeTakeId?: string;
  total: number;
}

export interface SetActiveTakeRequest {
  takeId: string;
}

export interface DeleteTakeResponse {
  success: boolean;
  message: string;
  newActiveTakeId?: string;
}

export interface TakeExportRequest {
  takeId: string;
  includeMetadata?: boolean;
}

export interface TakeExportResponse {
  success: boolean;
  exportPath?: string;
  message: string;
}

export interface TakeCleanupRequest {
  keepCount?: number;
  includeFailed?: boolean;
}

export interface TakeCleanupResponse {
  deletedCount: number;
  remainingCount: number;
  message: string;
}

export const takesApi = {
  /**
   * Create a new take for a shot
   */
  async createTake(
    projectId: string,
    shotId: string,
    request: CreateTakeRequest
  ): Promise<CreateTakeResponse> {
    return api.post<CreateTakeResponse>(
      `/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/takes`,
      request
    );
  },

  /**
   * List all takes for a shot
   */
  async listTakes(
    projectId: string,
    shotId: string,
    includeFailed = true
  ): Promise<TakeListResponse> {
    return api.get<TakeListResponse>(
      `/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/takes`,
      { includeFailed }
    );
  },

  /**
   * Get details for a specific take
   */
  async getTake(projectId: string, shotId: string, takeId: string): Promise<TakeMetadata> {
    return api.get<TakeMetadata>(
      `/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/takes/${takeId}`
    );
  },

  /**
   * Set the active take for a shot
   */
  async setActiveTake(
    projectId: string,
    shotId: string,
    takeId: string
  ): Promise<{ success: boolean; message: string }> {
    return api.put(`/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/active-take`, {
      takeId
    });
  },

  /**
   * Delete a take (soft delete)
   */
  async deleteTake(projectId: string, shotId: string, takeId: string): Promise<DeleteTakeResponse> {
    return api.delete<DeleteTakeResponse>(
      `/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/takes/${takeId}`
    );
  },

  /**
   * Export a take to the exports directory
   */
  async exportTake(
    projectId: string,
    shotId: string,
    request: TakeExportRequest
  ): Promise<TakeExportResponse> {
    return api.post<TakeExportResponse>(
      `/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/takes/export`,
      request
    );
  },

  /**
   * Clean up old takes for a shot
   */
  async cleanupTakes(
    projectId: string,
    shotId: string,
    request?: TakeCleanupRequest
  ): Promise<TakeCleanupResponse> {
    return api.post<TakeCleanupResponse>(
      `/takes/projects/${projectId}/shots/${encodeURIComponent(shotId)}/takes/cleanup`,
      request || { keepCount: 10 }
    );
  },

  /**
   * Get the thumbnail URL for a take
   */
  getThumbnailUrl(projectId: string, thumbnailPath?: string): string {
    if (!thumbnailPath) return '/placeholder-thumbnail.png';
    return `/api/v1/workspace/projects/${projectId}/files/${encodeURIComponent(thumbnailPath)}`;
  },

  /**
   * Get the media file URL for a take
   */
  getMediaUrl(projectId: string, filePath?: string): string {
    if (!filePath) return '';
    return `/api/v1/workspace/projects/${projectId}/files/${encodeURIComponent(filePath)}`;
  }
};
